CMakeLists.txt

```diff
@@ -69,6 +64,21 @@ string(REGEX MATCH "GNU" CMAKE_COMPILER_IS_GNU "${CMAKE_C_COMPILER_ID}")
 string(REGEX MATCH "IAR" CMAKE_COMPILER_IS_IAR "${CMAKE_C_COMPILER_ID}")
 string(REGEX MATCH "MSVC" CMAKE_COMPILER_IS_MSVC "${CMAKE_C_COMPILER_ID}")
 
+# Only attempt to auto-detect if the user has not explicitly
+# overriden the -D GEN_FILES.
+if(NOT DEFINED GEN_FILES)
+    option(GEN_FILES "Generate the auto-generated files as needed" ON)
+    execute_process(COMMAND git describe --exact-match HEAD
+                        OUTPUT_QUIET ERROR_QUIET RESULT_VARIABLE result)
```

Comment:
What if there's no git checkout (e.g. download via a zip)? Should we just assume a release in that case?

I still don't understand the advantage of this approach of detecting releases, comparing to changing the default from yes to no as part of the release process. We edit files to do a release anyway — in particular to include the generated files. So why not also edit one line in `Makefile` and one line in `CMakeLists.txt`?

Reply:
> What if there's no git checkout (e.g. download via a zip)? Should we just assume a release in that case?

When someone downloads the release as a tarball there are already manually involved, so I assume they can set the GEN_FILEs as desires, in addition to manually aquiring the files. 

But when they pull it as part of a scripted proccess, e.g through a CI or an intergration script, the presence of the annotated metadata is clear indicator that that is a release and we can automate the logic with it.


> I still don't understand the advantage of this approach of detecting releases, comparing to changing the default from yes to no as part of the release process. We edit files to do a release anyway — in particular to include the generated files. So why not also edit one line in Makefile and one line in CMakeLists.txt?

If the expectation is for the release builder to be manually involved then they could simply pass the `GEN_FILES=OFF/0` as part of their workflow.

The original ask is:
> The goal of this issue is to completely turn off the generation of these files in a release. 

This cannot be done implicitly implicitly since there is no release target. That means for the default to apply the tools need to know the user is building a release. This could be achieved by:

- Adding yet another new build flag that needs to be passed in during the configure command to indicate a release. I see no benefit to that since the same can be achieved by setting `GEN_FILES=OFF`
- Using the annotated metadata to assertain with confidence that this is a relase and disable automatic file generation while still allowing complete override when the `GEN_FILES` is passed through the command line.

I though that the latter is a big more elegant. If you feel that it makes no sense, I am open to switching the first approach or other suggestions.

Reply:
> When someone downloads the release as a tarball there are already manually involved, so I assume they can set the GEN_FILEs as desires, in addition to manually aquiring the files.

Uh? No. They just download the source. For example using the “Download ZIP” button on https://github.com/Mbed-TLS/mbedtls.

> But when they pull it as part of a scripted proccess, e.g through a CI or an intergration script, the presence of the annotated metadata is clear indicator that that is a release and we can automate the logic with it.

What metadata?

> > The original ask is:
>
> The goal of this issue is to completely turn off the generation of these files in a release.

Yes, and what you're implementing here is a weirdly roundabout way of doing it, compared to what I expected when I wrote this, which is to replace `GEN_FILES = ON` by `GEN_FILES = OFF` (or whatever) as part of our release process.

> We edit files to do a release anyway — in particular to include the generated files. So why not also edit one line in Makefile and one line in CMakeLists.txt?
>
> If the expectation is for the release builder to be manually involved then they could simply pass the GEN_FILES=OFF/0 as part of their workflow.

Who's “they” here? The release builder is us.